core/controllers/question_editor.py

```diff
@@ -115,24 +115,43 @@ class QuestionSkillLinkHandler(base.BaseHandler):
     """A handler for linking and unlinking questions to or from a skill."""
 
     @acl_decorators.can_manage_question_skill_status
-    def put(self, question_id, skill_id):
-        """Updates the difficulty of the question with respect to the given
-        skill.
+    def put(self, question_id):
+        """Updates the QuestionSkillLink models with respect to the given
+        question.
         """
-        new_difficulty = float(self.payload.get('new_difficulty'))
-
-        question_services.update_question_skill_link_difficulty(
-            question_id, skill_id, new_difficulty)
+        if self.payload.get('action') == 'update_difficulty':
+            new_difficulty = self.payload.get('new_difficulty')
+            skill_id = self.payload.get('skill_id')
+            if skill_id is None:
+                raise self.InvalidInputException(
+                    'The \'skill_id\' field is missing in payload')
+            if new_difficulty is None:
+                raise self.InvalidInputException(
+                    'The \'new_difficulty\' field is missing in payload')
+            question_services.update_question_skill_link_difficulty(
+                question_id, skill_id, float(new_difficulty))
+        elif self.payload.get('action') == 'edit_links':
+            difficulty = self.payload.get('difficulty')
+            skill_ids_task_list = self.payload.get('skill_ids_task_list')
+            if skill_ids_task_list is None:
+                raise self.InvalidInputException(
+                    'Missing fields \'skill_ids_task_list\'in payload')
+
+            for task_dict in skill_ids_task_list:
+                if task_dict['task'] == 'remove':
+                    question_services.delete_question_skill_link(
+                        self.user_id, question_id, task_dict['id'])
+                elif task_dict['task'] == 'add':
```

Comment:
There should be an "else raise exception" after this -- remember, everything in the input should be untrusted by default.

Could you please go through this handler carefully and ensure that every possible invalid input is caught?

Reply:
Done, and added another exception.